pkg/smokescreen/conntrack/conn_tracker_test.go

```diff
@@ -51,3 +51,55 @@ func NewTestTracker(idle time.Duration) *Tracker {
 
 	return NewTracker(idle, &statsd.NoOpClient{}, logrus.New(), sd, nil)
 }
+
+func TestConnSuccessRateTracker(t *testing.T) {
+	type record struct {
+		host    string
+		success bool
+	}
+
+	var testCases = []struct {
+		name         string
+		additions    []record
+		waitTime     time.Duration
+		expectedRate float64
+		totalConns   int
+	}{
+		{"fifty-percent-success", []record{{"foo.com", true}, {"bar.com", false}}, 1 * time.Second, 50.0, 2},
+		{"no-records", []record{}, 0 * time.Second, 100.0, 0},
+		// If we wait beyond the calculation window, we should have no records and return the default 100% rate
+		{"expire-records", []record{{"foo.com", true}, {"bar.com", false}}, 3 * time.Second, 100.0, 0},
+		// Only the most recent record for a host is used in the computation
+		{"dedup-hosts", []record{{"foo.com", false}, {"bar.com", false}, {"foo.com", true}}, 1 * time.Second, 50.0, 2},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert := assert.New(t)
+
+			sd := atomic.Value{}
+			sd.Store(false)
+			tracker := NewTracker(time.Second, &statsd.NoOpClient{}, logrus.New(), sd, StartNewConnSuccessRateTracker(500*time.Millisecond, 2*time.Second))
+
+			for _, record := range tc.additions {
+				tracker.RecordAttempt(record.host, record.success)
+			}
+
+			time.Sleep(tc.waitTime)
```

Comment:
I gave it a try, but since the cache library (https://github.com/patrickmn/go-cache/blob/master/cache_test.go) doesn't support injecting a clock it doesn't give us a great way to test cache expiry.